File: Semester-2/Graphs/Assignment01/Assignment01/graph.cpp
#pragma warning(disable : 4996)

#include "graph.h"
#include <string>
#include <exception>
#include <fstream>
#include <stdio.h>
#include <ctime>
#include <random>


#define NoEdge "There is no edge between the specified vertices"
#define DuplicateEdge "Duplicate Edge"
#define DuplcateVertex "Duplicate Vertex"
#define NoVertex "There is no vertex with the specified value"
#define InvalidNumberOfVertices "The number of vertices must be a positive number"
#define InvalidNumberOfEdges "The number of edges must be a positive number and must be less then (noOfVertices * noOfVertices)"
#define NoLoopsAllowed "You can not add a loop edge"


std::string getID(int vertex0, int vertex1)
{
    return std::to_string(vertex0) + "-" + std::to_string(vertex1);
}

Graph::Graph()
{
}

Graph::Graph(int noOfVertices)
{
    if (noOfVertices < 0)
    {
        throw std::exception(InvalidNumberOfVertices);
    }

    for (int i = 0; i < noOfVertices; i++)
    {
        this->add_vertex(i);
    }
}

Graph::Graph(const Graph& graph)
{
    this->vertices = graph.vertices;
    this->inbound = graph.inbound;
    this->outbound = graph.outbound;

    this->edges = graph.edges;
}

bool Graph::is_vertex(int vertex) const
{
    return this->vertices.find(vertex) != this->vertices.end();
}

bool Graph::is_edge(int vertex0, int vertex1) const
{
    if (!this->is_vertex(vertex0) || !this->is_vertex(vertex1))
    {
        throw std::exception(NoVertex);
    }

    return this->outbound.at(vertex0).find(vertex1) != this->outbound.at(vertex0).end();
}

int Graph::count_vertices() const
{
    return this->vertices.size();
}

int Graph::count_edges() const
{
    return this->edges.size();
}

int Graph::get_in_degree(int vertex) const
{
    if (!this->is_vertex(vertex))
    {
        throw std::exception(NoVertex);
    }

    return (int)this->inbound.at(vertex).size();
}

int Graph::get_out_degree(int vertex) const
{
    if (!this->is_vertex(vertex))
    {
        throw std::exception(NoVertex);
    }

    return (int)this->outbound.at(vertex).size();
}


void Graph::set_edge_data(int vertex0, int vertex1, int value)
{
    if (!this->is_vertex(vertex0) || !this->is_vertex(vertex1))
    {
        throw std::exception(NoVertex);
    }

    if (!this->is_edge(vertex0, vertex1))
    {
        throw std::exception(NoEdge);
    }

    this->edges[getID(vertex0, vertex1)] = value;
}

int Graph::get_edge_data(int vertex0, int vertex1) const
{
    if (!this->is_vertex(vertex0) || !this->is_vertex(vertex1))
    {
        throw std::exception(NoVertex);
    }

    if (!this->is_edge(vertex0, vertex1))
    {
        throw std::exception(NoEdge);
    }

    return this->edges.at(getID(vertex0, vertex1));
}

void Graph::add_edge(int vertex0, int vertex1)
{
    if (!this->is_vertex(vertex0) || !this->is_vertex(vertex1))
    {
        throw std::exception(NoVertex);
    }

    if (this->is_edge(vertex0, vertex1))
    {
        throw std::exception(DuplicateEdge);
    }

    this->outbound[vertex0].insert(vertex1);
    this->inbound[vertex1].insert(vertex0);
    this->set_edge_data(vertex0, vertex1, 0);
}

void Graph::remove_edge(int vertex0, int vertex1)
{
    if (!this->is_vertex(vertex0) || !this->is_vertex(vertex1))
    {
        throw std::exception(NoVertex);
    }

    if (!this->is_edge(vertex0, vertex1))
    {
        throw std::exception(NoEdge);
    }

    this->outbound[vertex0].erase(vertex1);
    this->inbound[vertex1].erase(vertex0);
    this->edges.erase(getID(vertex0, vertex1));
}

void Graph::add_vertex(int vertex)
{
    if (this->is_vertex(vertex))
    {
        throw std::exception(DuplcateVertex);
    }

    this->vertices.insert(vertex);
    this->inbound[vertex] = std::unordered_set<int>();
    this->outbound[vertex] = std::unordered_set<int>();
}

void Graph::remove_vertex(int vertex)
{
    if (!this->is_vertex(vertex))
    {
        throw std::exception(NoVertex);
    }

    for (auto to : this->outbound[vertex])
    {
        this->remove_edge(vertex, to);
    }

    for (auto from : this->inbound[vertex])
    {
        this->remove_edge(from, vertex);
    }

    this->vertices.erase(vertex);
    this->inbound.erase(vertex);
    this->outbound.erase(vertex);
}


std::unordered_set<int>::iterator Graph::vertices_begin() const
{
    return this->vertices.begin();
}

std::unordered_set<int>::iterator Graph::vertices_end() const
{
    return this->vertices.end();
}

std::unordered_set<int>::iterator Graph::in_begin(int vertex) const
{
    return this->inbound.at(vertex).begin();
}

std::unordered_set<int>::iterator Graph::in_end(int vertex) const
{
    return this->inbound.at(vertex).end();
}

std::unordered_set<int>::iterator Graph::out_begin(int vertex) const
{
    return this->outbound.at(vertex).begin();
}

std::unordered_set<int>::iterator Graph::out_end(int vertex) const
{
    return this->outbound.at(vertex).end();
}

Graph::~Graph()
{

}



Graph readGraphFromFile(std::string filePath)
{
    std::ifstream fin;
    fin.open(filePath);

    int n, m;

    fin >> n >> m;

    Graph g{ n };

    int vertex0, vertex1, value;

    for (int i = 0; i < m; i++)
    {
        fin >> vertex0 >> vertex1 >> value;
        g.add_edge(vertex0, vertex1);
        g.set_edge_data(vertex0, vertex1, value);
    }

    fin.close();
    return g;
}


void writeGraphToFile(Graph g, std::string filePath)
{
    std::ofstream fout;
    fout.open(filePath);
    fout << g.count_vertices() << ' ' << g.count_edges() << '\n' << g;
    fout.close();
}

Graph createRandomGraph(int vertices, int edges)
{
    if (vertices < 0)
    {
        throw std::exception(InvalidNumberOfVertices);
    }

    if (edges < 0 || vertices * vertices < edges)
    {
        throw std::exception(InvalidNumberOfEdges);
    }
    
    std::random_device dev;
    std::mt19937 rng(dev());
    std::uniform_int_distribution<std::mt19937::result_type> random_cost(0, 200);
    std::uniform_int_distribution<std::mt19937::result_type> random_vertex(0, vertices - 1);

    Graph g{ vertices };
    
    for (int i = 0; i < edges; i++)
    {
        try
        {
            int vertex0 = random_vertex(rng);
            int vertex1 = random_vertex(rng);
            int cost = random_cost(rng);

            g.add_edge(vertex0, vertex1);
            g.set_edge_data(vertex0, vertex1, cost);
        }
        catch (...)
        {
            i--;
        }
    }

    return g;
}

File: Semester-2/Graphs/Assignment01/Python/graph.py
import copy
from random import randrange

NoEdge = "There is no edge between the specified vertices"
DuplicateEdge = "Duplicate Edge"
DuplicateVertex = "Duplicate Vertex"
NoVertex  = "There is no vertex with the specified value"
InvalidNumberOfVertices = "The number of vertices must be a positive number"
InvalidNumberOfEdges = "The number of edges must be a positive number and must be less then (noOfVertices * noOfVertices)"

class Graph:

    def __init__(self, n = 0):

        if n < 0:
            raise ValueError(InvalidNumberOfVertices)

        self._vertices = set()
        self._inbound = dict()
        self._outbound = dict()
        self._edges = dict()

        
        for i in range(n):
            self.add_vertex(i)


    def is_vertex(self, vertex):
        return vertex in self._vertices


    def is_edge(self, vertex0, vertex1):

        if vertex0 not in self._vertices or vertex1 not in self._vertices:
            raise ValueError(NoVertex)

        return (vertex0, vertex1) in self._edges.keys()


    def count_vertices(self):
        return len(self._vertices)


    def count_edges(self):
        return len(self._edges)


    def get_in_degree(self, vertex):
        if vertex not in self._vertices:
            raise ValueError(NoVertex)

        return len(self._inbound[vertex])


    def get_out_degree(self, vertex):
        if vertex not in self._vertices:
            raise ValueError(NoVertex)

        return len(self._outbound[vertex])


    def set_edge_data(self, vertex0, vertex1, value):
        if vertex0 not in self._vertices or vertex1 not in self._vertices:
            raise ValueError(NoVertex)

        if (vertex0, vertex1) not in self._edges:
            raise ValueError(NoEdge)

        self._edges[(vertex0, vertex1)] = value


    def get_edge_data(self, vertex0, vertex1):
        if vertex0 not in self._vertices or vertex1 not in self._vertices:
            raise ValueError(NoVertex)

        if (vertex0, vertex1) not in self._edges:
            raise ValueError(NoEdge)

        return self._edges[(vertex0, vertex1)]


    def add_edge(self, vertex0, vertex1):
        if vertex0 not in self._vertices or vertex1 not in self._vertices:
            raise ValueError(NoVertex)

        if (vertex0, vertex1) in self._edges:
            raise ValueError(DuplicateEdge)

        self._outbound[vertex0].add(vertex1)
        self._inbound[vertex1].add(vertex0)
        self._edges[(vertex0, vertex1)] = 0


    def remove_edge(self, vertex0, vertex1):
        if vertex0 not in self._vertices or vertex1 not in self._vertices:
            raise ValueError(NoVertex)

        if (vertex0, vertex1) not in self._edges:
            raise ValueError(NoEdge)

        self._outbound[vertex0].remove(vertex1)
        self._inbound[vertex1].remove(vertex0)
        del self._edges[(vertex0, vertex1)]


    def add_vertex(self, vertex):
        if vertex in self._vertices:
            raise ValueError(DuplicateVertex)

        self._vertices.add(vertex)
        self._inbound[vertex] = set()
        self._outbound[vertex] = set()


    def remove_vertex(self, vertex):
        if vertex not in self._vertices:
            raise ValueError(NoVertex)

        remove = []

        for neighbor in self._outbound[vertex]:
            remove.append((vertex, neighbor))

        for neighbor in self._inbound[vertex]:
            remove.append((neighbor, vertex))
        
        for edge in remove:
            self.remove_edge(edge[0], edge[1])

        
        del self._inbound[vertex]
        del self._outbound[vertex]
        self._vertices.remove(vertex)


    def vertices_iterator(self):
        for vertex in self._vertices:
            yield vertex


    def inbound_iterator(self, vertex):
        if vertex not in self._vertices:
            return ValueError(NoVertex)

        for n in self._inbound[vertex]:
            yield n


    def outbound_iterator(self, vertex):
        if vertex not in self._vertices:
            return ValueError(NoVertex)

        for n in self._outbound[vertex]:
            yield n


    def get_copy(self):
        return copy.deepcopy(self)


    def __str__(self):

        string = ""

        for vertex in self._vertices:
            for y in self._outbound[vertex]:
                string += f"{vertex} {y} {self.get_edge_data(vertex, y)}\n"

        return string


def read_from_file(path):

    with open(path, "r") as file:
        lines = file.readlines()
        line = lines[0].split()
        vertices = int(line[0])
        edges = int(line[1])
       
        g = Graph(vertices)

        for i in range(1, edges + 1):
            line = lines[i].split()
            vertex0 = int(line[0])
            vertex1 = int(line[1])
            cost = int(line[2])
            g.add_edge(vertex0, vertex1)
            g.set_edge_data(vertex0, vertex1, cost)

    return g

def write_to_file(graph: Graph, path):
    with open(path, "w") as file:
        file.write(f"{graph.count_vertices()} {graph.count_edges()}\n")

        for x in graph.vertices_iterator():
            for y in graph.outbound_iterator(x):
                file.write(f"{x} {y} {graph.get_edge_data(x, y)}\n")


def generate_graph(n, m):

    if m < 0 or n * n < m:
        raise ValueError(InvalidNumberOfEdges)

    g = Graph(n)

    i = 0
    while i < m:
        x = randrange(n)
        y = randrange(n)
        cost = randrange(200)

        try:
            g.add_edge(x, y)
            g.set_edge_data(x, y, cost)
        except:
            i -= 1

        i += 1

    return g